assets/shared/helpers/player.js

```diff
@@ -0,0 +1,557 @@
+ */
+class Player {
+	/**
+	 * Player constructor.
+	 *
+	 * @param {HTMLVideoElement|HTMLIFrameElement} element The player element.
+	 */
+	constructor( element ) {
+		this.playerPromise = null;
+		this.type = null;
+		this.element = element;
+
+		try {
+			this.setType();
+		} catch ( e ) {
+			// eslint-disable-next-line no-console -- We want to expose the element with problem.
+			console.error( e, element );
+		}
+	}
+
+	/**
+	 * Set the player type.
+	 *
+	 * @throws Will throw an error if the video type is not found.
+	 */
+	setType() {
+		if ( this.element instanceof window.HTMLVideoElement ) {
+			this.type = VIDEO_TYPE;
+		} else if ( this.element instanceof window.HTMLIFrameElement ) {
+			this.type = Object.entries( players ).find(
+				( [ , p ] ) =>
+					p.embedPattern && this.element.src?.match( p.embedPattern )
+			)?.[ 0 ];
+		}
+
+		if ( ! this.type ) {
+			throw new Error( 'Video type not found' );
+		}
+	}
+
+	/**
+	 * Get the video player.
+	 *
+	 * @return {Promise<Object|HTMLVideoElement|HTMLIFrameElement>} The video player through a promise.
+	 */
+	getPlayer() {
+		if ( ! this.playerPromise ) {
+			this.playerPromise =
+				players[ this.type ]?.initializePlayer( this.element ) ||
+				// A promise that never resolves if it doesn't exist.
+				Promise.reject( new Error( 'Failed getting the player' ) );
+		}
+
+		return this.playerPromise;
+	}
+
+	/**
+	 * Get the video duration.
+	 *
+	 * @return {Promise<number>} The duration of the video in seconds through a promise.
+	 */
+	getDuration() {
+		return this.getPlayer().then( ( player ) =>
+			players[ this.type ].getDuration( player )
+		);
+	}
+
+	/**
+	 * Set the video to a current time.
+	 *
+	 * @param {number} seconds The video time in seconds to set.
+	 *
+	 * @return {Promise} A promise that resolves if the video was set to a current time successfully.
+	 */
+	setCurrentTime( seconds ) {
+		return this.getPlayer().then( ( player ) =>
+			players[ this.type ].setCurrentTime( player, seconds )
+		);
+	}
+
+	/**
+	 * Play the video.
+	 *
+	 * @return {Promise} A promise that resolves if the video was played successfully.
+	 */
+	play() {
+		return this.getPlayer().then( ( player ) =>
+			players[ this.type ].play( player )
+		);
+	}
+
+	/**
+	 * Pause the video.
+	 *
+	 * @return {Promise} A promise that resolves if the video was paused successfully.
+	 */
+	pause() {
+		return this.getPlayer().then( ( player ) =>
+			players[ this.type ].pause( player )
+		);
+	}
+
+	/**
+	 * Add an event listener to the player.
+	 *
+	 * @param {string}   eventName Event name (supported: `timeupdate`).
+	 * @param {Function} callback  Listener callback.
+	 *
+	 * @throws Will throw an error if the event is not supported.
+	 *
+	 * @return {Promise<Function>} The function to unsubscribe the event through a promise.
+	 */
+	on( eventName, callback ) {
+		// Check supported events.
+		if ( eventName !== 'timeupdate' ) {
```

Comment:
Ahhh this is where we check. So, as mentioned above, it still seems a little weird to me that the other functions are written to only support `timeupdate` but don't explicitly prevent other events. Not sure whether we want to duplicate this check though? Seems like it might make it a little more clear.

Alternatively, maybe if the players didn't just return the current time, that would make more sense? Like, if they return a more generic event object that could be used by different events, and this function can ensure that only the timestamp is returned for the `timeupdate` event?

But I'm not really sure. It doesn't seem like a huge problem. The code just looks a little strange 🙂

Reply:
I got your point!

> Alternatively, maybe if the players didn't just return the current time, that would make more sense? Like, if they return a more generic event object that could be used by different events, and this function can ensure that only the timestamp is returned for the timeupdate event?

I think the problem is that depending on the event we need to handle it in a different way to make all the players have the same signature in our class.

But I had an idea to have the `on` just on our "public" method. And in the players, we can have a different type of function, like `onTimeupdate`, so it doesn't get weird anymore. Does it make sense?

Reply:
That sounds good to me!

Reply:
Updated here: https://github.com/Automattic/sensei/pull/5377/commits/c277033007a08b7569e41fef359f541ad3fc7e99